core/src/main/java/org/apache/accumulo/core/iterators/user/SeekingFilter.java

```diff
@@ -0,0 +1,221 @@
+    negate = Boolean.parseBoolean(options.get(NEGATE));
+  }
+
+  @Override
+  public SortedKeyValueIterator<Key,Value> deepCopy(IteratorEnvironment env) {
+    SeekingFilter newInstance;
+    try {
+      newInstance = this.getClass().newInstance();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    newInstance.setSource(getSource().deepCopy(env));
+    newInstance.negate = negate;
+    return newInstance;
+  }
+
+  protected void findTop() throws IOException {
+    SortedKeyValueIterator<Key,Value> src = getSource();
+    // advance could be null if we've just been seeked
+    if (src.hasTop() && advance != null) {
+      advanceSource(src, advance);
+    }
+    advance = null;
+    while (src.hasTop() && !advancedPastSeek) {
+      if (src.getTopKey().isDeleted()) {
+        // as per. o.a.a.core.iterators.Filter, deleted keys always pass through the filter.
+        advance = AdvanceResult.NEXT;
+        return;
+      }
+      FilterResult f = filter(src.getTopKey(), src.getTopValue());
+      if (log.isTraceEnabled()) {
+        log.trace("Filtered: " + src.getTopKey() + " result == " + f + " hint == " + getNextKeyHint(src.getTopKey(), src.getTopValue()));
+      }
+      if (f.accept != negate) {
+        // advance will be processed next time findTop is called
+        advance = f.advance;
+        break;
+      } else {
+        advanceSource(src, f.advance);
+      }
+    }
+  }
+
+  private void advanceSource(SortedKeyValueIterator<Key,Value> src, AdvanceResult adv) throws IOException {
+    Key topKey = src.getTopKey();
+    Range advRange = null;
+    switch (adv) {
+      case NEXT:
+        src.next();
+        return;
+      case NEXT_CQ:
+        advRange = new Range(topKey.followingKey(PartialKey.ROW_COLFAM_COLQUAL), null);
+        break;
+      case NEXT_CF:
+        advRange = new Range(topKey.followingKey(PartialKey.ROW_COLFAM), null);
+        break;
+      case NEXT_ROW:
+        advRange = new Range(topKey.followingKey(PartialKey.ROW), null);
+        break;
+      case USE_HINT:
+        Value topVal = src.getTopValue();
+        Key hintKey = getNextKeyHint(topKey, topVal);
+        if (hintKey != null && hintKey.compareTo(topKey) > 0) {
+          advRange = new Range(hintKey, null);
+        } else {
+          String msg = "Filter returned USE_HINT for " + topKey + " but invalid hint: " + hintKey;
+          throw new IOException(msg);
+        }
+        break;
+    }
+    if (advRange == null) {
+      // Should never get here. Just a safeguard in case somebody adds a new type of AdvanceRange and forgets to handle it here.
+      throw new IOException("Unable to determine range to advance to for AdvanceResult " + adv);
+    }
+    advRange = advRange.clip(seekRange, true);
+    if (advRange == null) {
+      // the advanced range is outside the seek range. the source is exhausted.
+      advancedPastSeek = true;
+    } else {
+      src.seek(advRange, columnFamilies, inclusive);
```

Comment:
Yeah, I've seen that pattern in eg. RowFilter. It would be interesting but I think I'll leave it for another day.